### PR TITLE
perf: Move DependencyPropertyGenerator to be incremental

### LINKED ISSUE #9914: Move DependencyPropertyGenerator to be incremental
Currently, the internal generators project runs only on Roslyn, and we can easily consume newer version of the compiler. I think we should try moving `DependencyPropertyGenerator` to be incremental and make use of `ForAttributeWithMetadataName` API introduced in https://github.com/dotnet/roslyn/pull/61293

Note: that this issue is easier to tackle than https://github.com/unoplatform/uno/issues/8458 since it's concerned only by internal generators that are not shipped or usable by Uno consumers.

### LINKED ISSUE #9940: Possible bug in DependencyPropertyGenerator
### Current behavior

https://github.com/unoplatform/uno/blob/c2b814cdecf24358d45348370c63336faf8a1456/src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs#L478-L481

I think `GetFirstMethodWithName` should never return anything with the `()` appended to the method name.

I'll need to validate first though, but just discovered while re-writing this generator to be incremental.

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

_No response_

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #9914, closes #9940

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 1]
```diff
@@ -5,527 +5,510 @@
...
 
 namespace Uno.UI.SourceGenerators.DependencyObject
 {
+	// TODO: We are not respecting IsValidPlatform now.
 	[Generator]
-	public class DependencyPropertyGenerator : ISourceGenerator
+	public class DependencyPropertyGenerator : IIncrementalGenerator
 	{
-		public void Initialize(GeneratorInitializationContext context)
+		private sealed class ContainingTypeEqualityComparer : IEqualityComparer<ISymbol?>
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:20]: This is unused. I'll delete

**Review by MartinZikmund — APPROVED** [review]

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 2]
```diff
@@ -5,527 +5,498 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Uno.Extensions;
 using Uno.UI.SourceGenerators.Helpers;
+using Uno.UI.SourceGenerators.Internal.Extensions;
 using Uno.UI.SourceGenerators.XamlGenerator;
 
 namespace Uno.UI.SourceGenerators.DependencyObject
 {
+	// TODO: We are not respecting IsValidPlatform now.
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:17]: Maybe there is no need to do this check. Things are passing and working very well.

I think the reason is that now `Uno.UI.SourceGenerators.Internal` included in less projects, as opposed to when this generator was in Uno.UI.SourceGenerators project which run on much more projects.

Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:17]: Indeed. For this particular generator, it was only needed to avoid running it when the requested attribute was not present. Inside of Uno, it's always available.

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 3]
```diff
@@ -5,527 +5,498 @@
...
-
-			public override void VisitNamespace(INamespaceSymbol symbol)
+				var fieldOrPropertyData = combined.Left;
+				var dependencyObjectSymbol = combined.Right;
+				var containingType = fieldOrPropertyData.FieldOrPropertySymbol.ContainingType;
+				return containingType.TypeKind == TypeKind.Class &&
+					(containingType.IsStatic || containingType.AllInterfaces.Any(t => SymbolEqualityComparer.Default.Equals(t, dependencyObjectSymbol)));
+			});
+
+			var groupedByContainingProvider = filteredAttributedSymbolsProvider.Select((x, _) => x.Left).GroupBy(data => data.FieldOrPropertySymbol.ContainingType, SymbolEqualityComparer.Default);
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:55]:
```suggestion
			var groupedByContainingProvider = filteredAttributedSymbolsProvider
			   .Select((x, _) => x.Left)
			   .GroupBy(data => data.FieldOrPropertySymbol.ContainingType, SymbolEqualityComparer.Default);
```

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 4]
```diff
@@ -5,527 +5,498 @@
...
+			var finalProvider = groupedByContainingProvider.Combine(context.CompilationProvider.Select((c, _) => c.GetTypeByMetadataName("Windows.UI.Xaml.DependencyPropertyChangedEventArgs")));
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:56]:
```suggestion
			var finalProvider = groupedByContainingProvider
			   .Combine(context.CompilationProvider
			      .Select((c, _) => 
			         c.GetTypeByMetadataName("Windows.UI.Xaml.DependencyPropertyChangedEventArgs")));
```

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 5]
```diff
@@ -5,527 +5,498 @@
...
+			context.RegisterSourceOutput(finalProvider, static (context, item) =>
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:57]: Let's exact the lambda to another method, this is getting hard to follow :)

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 6]
```diff
@@ -5,527 +5,498 @@
...
-				VisitNamespace(symbol.GlobalNamespace);
-			}
-
-			public override void VisitNamespace(INamespaceSymbol symbol)
+				var fieldOrPropertyData = combined.Left;
+				var dependencyObjectSymbol = combined.Right;
+				var containingType = fieldOrPropertyData.FieldOrPropertySymbol.ContainingType;
+				return containingType.TypeKind == TypeKind.Class &&
+					(containingType.IsStatic || containingType.AllInterfaces.Any(t => SymbolEqualityComparer.Default.Equals(t, dependencyObjectSymbol)));
+			});
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:53]:
```suggestion
				var isDependencyObject = containingType.AllInterfaces
				   .Any(t => SymbolEqualityComparer.Default.Equals(t, dependencyObjectSymbol));
				
				return containingType.TypeKind == TypeKind.Class &&
					(containingType.IsStatic || isDependencyObject );
			});
```

**Review by jeromelaban — COMMENTED** [review]
Looking good! Let's make it a bit more readable with shorter lines.

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs (hunk 2 above)
Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:17]:
```suggestion
```

**Review by Youssef1313 — COMMENTED** [review]
The generator now is hard refactored. The idea is that generator pipeline should be dealing with types that are cache friendly.

Symbols are not cache friendly, meaning that two symbols from two different compilations are not equal.

So, we build a data model for the data we want from the symbol.
If all we need from a symbol is a fully qualified name, we add that to the data model. This allows Roslyn to make use of caching mechanisms because the fully qualified name as a string can be equal across different compilations (as opposed to comparing the ISymbol).

File: src/Uno.UI/UI/Xaml/Controls/Grid/RowDefinition.cs [hunk 7]
```diff
@@ -51,8 +51,6 @@ public double MinHeight
 			set => SetMinHeightValue(value);
 		}
 
-		private static GridLength GetMaxHeightDefaultValue() => GridLengthHelper.OneStar;
-
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/RowDefinition.cs:55]: Nice catch

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 8]
```diff
@@ -2,530 +2,589 @@
...
-	public class DependencyPropertyGenerator : ISourceGenerator
+	public class DependencyPropertyGenerator : IIncrementalGenerator
 	{
-		public void Initialize(GeneratorInitializationContext context)
-		{
-		}
+		private static SymbolDisplayFormat _fullyQualifiedWithoutGlobal = SymbolDisplayFormat.FullyQualifiedFormat.WithGlobalNamespaceStyle(SymbolDisplayGlobalNamespaceStyle.Omitted);
 
-		public void Execute(GeneratorExecutionContext context)
+		private readonly struct AttachedData
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:23]:
```suggestion
		private readonly struct AttachedPropertyData
```

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 9]
```diff
@@ -2,530 +2,589 @@
...
+				HasSetMethodSymbol = setMethodSymbol is not null;
+				PropertyTypeFullyQualifiedName = getMethodSymbol?.ReturnType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+				PropertyTargetFullyQualifiedName = getMethodSymbol?.Parameters[0].Type.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
 			}
+		}
 
-			public override void VisitModule(IModuleSymbol symbol)
-			{
-				_context.CancellationToken.ThrowIfCancellationRequested();
+		private readonly struct NonAttachedData
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:59]:
```suggestion
		private readonly struct PropertyData
```

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 10]
```diff
@@ -2,530 +2,589 @@
...
+				
+				PropertyHasGetter = propertySymbol?.GetMethod is not null;
+				PropertyHasSetter = propertySymbol?.SetMethod is not null;
 			}
+		}
 
-			private void GenerateAttachedProperty(IndentedStringBuilder builder, INamedTypeSymbol ownerType, ISymbol memberSymbol, AttributeData attribute, Dictionary<INamedTypeSymbol, List<string>> backingFieldStatements)
-			{
-				var propertyName = memberSymbol.Name.TrimEnd("Property", StringComparison.Ordinal);
+		private readonly struct FieldOrPropertyData
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:85]:
```suggestion
		private readonly struct GenerationCandidateData
```

**Review by jeromelaban — COMMENTED** [review]
I'm suggesting new names that I think would match better the intent of the new structures

File: src/Uno.UI/UI/Xaml/Controls/Grid/RowDefinition.cs (hunk 7 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/Grid/RowDefinition.cs:55]: This was caught by the fix for https://github.com/unoplatform/uno/issues/9940 🎉

File: src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs [hunk 11]
```diff
@@ -2,530 +2,589 @@
...
-				=> SymbolEqualityComparer.Default.Equals(m?.Parameters.FirstOrDefault()?.Type, _dependencyPropertyChangedEventArgsSymbol);
+		public void Initialize(IncrementalGeneratorInitializationContext context)
+		{
+			var attributedSymbolsProvider = context.SyntaxProvider.ForAttributeWithMetadataName(
+				"Uno.UI.Xaml.GeneratedDependencyPropertyAttribute",
+				static (node, _) => node.IsKind(SyntaxKind.PropertyDeclaration),
+				static (context, token) =>
+				{
+					var attribute = context.Attributes[0];
+					return new GenerationCandidateData(context.TargetSymbol, attribute);
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:191]: Do you know how roslyn compute the equality for `GenerationCandidateData`? Is it done through reflection?

Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators.Internal/DependencyObject/DependencyPropertyGenerator.cs:191]: I'm not sure. But it's quite possible that structs are compared through reflection by default.

Let me use a `record`

**Review by jeromelaban — APPROVED** [review]
